File: java2/src/연속펄스부분수열의합/연속펄스부분수열의합.java
package 연속펄스부분수열의합;

class 연속펄스부분수열의합  {

    public static int solution(int[] sequence) {

        boolean pivot = true;
        int[] s1 = new int[sequence.length];
        for (int i = 0; i < sequence.length; ++i) {
            if (pivot) {
                s1[i] = sequence[i];
                pivot = false;
            } else {
                s1[i] = -1 * sequence[i];
                pivot = true;
            }
        }
        int current_max = s1[0];
        int global_max = s1[0];

        int current_max2 = sequence[0] * -1;
        int global_max2 = sequence[0] * -1;

        for (int i = 1; i < s1.length; ++i) {
            int s1Index = s1[i];
            int s2Index = s1[i] * -1;

            current_max = Math.max(s1Index, current_max + s1Index);
            current_max2 = Math.max(s2Index, current_max2 + s2Index);

            global_max = Math.max(global_max, current_max);
            global_max2 = Math.max(global_max2, current_max2);
        }

        return Math.max(global_max, global_max2);
    }


    public static void main(String[] args) {
        int[] sequence = {2, 3, -6, 1, 3, -1, 2, 4};
        System.out.println(solution(sequence));

    }
}

// https://school.programmers.co.kr/learn/courses/30/lessons/152995

File: python/algorithm/everything_algorithm/algorithm_study/programmers/level3/연속펄스부분수열의합/연속펄스부분수열의합.py
def solution(sequence):
    answer = 0

    sequence_r1 = []
    sequence_r2 = []
    pivot = True
    for i in range(len(sequence)):
        if pivot:
            sequence_r1.append(sequence[i] * 1)
            sequence_r2.append(sequence[i] * -1)
            pivot = False
        else:
            sequence_r1.append(sequence[i] * -1)
            sequence_r2.append(sequence[i] * 1)
            pivot = True

    start, end, list_length = 0, 0, len(sequence)
    while start < list_length:
        answer = max(answer, sum(sequence_r1[start:end + 1]), sum(sequence_r2[start:end + 1]))
        end += 1

        if end >= list_length - 1:
            start += 1
            end = start

    return answer

print(solution([2, 3, -6, 1, 3, -1, 2, 4]))
# 10

# https://school.programmers.co.kr/learn/courses/30/lessons/161988